embassy-extras/src/peripheral.rs

```diff
@@ -52,22 +157,61 @@ impl<S: PeripheralState> PeripheralMutex<S> {
         this.irq_setup_done = true;
     }
 
-    pub fn with<R>(self: Pin<&mut Self>, f: impl FnOnce(&mut S, &mut S::Interrupt) -> R) -> R {
+    pub fn with<R>(self: Pin<&mut Self>, f: impl FnOnce(&mut S) -> R) -> R {
         let this = unsafe { self.get_unchecked_mut() };
 
+        let was_enabled = this.irq.is_enabled();
         this.irq.disable();
 
         // Safety: it's OK to get a &mut to the state, since the irq is disabled.
         let state = unsafe { &mut *this.state.get() };
-        let r = f(state, &mut this.irq);
+        let r = f(state);
 
-        this.irq.enable();
+        if was_enabled {
+            this.irq.enable();
+        }
 
         r
     }
+
+    /// Enables the wrapped interrupt.
+    pub fn enable(&self) {
```

Comment:
If we require the irq to be permanently enabled for the lifetime of the PeripheralMutex (see above comment) then it'd make sense to not expoes enable, disable, is_enabled.